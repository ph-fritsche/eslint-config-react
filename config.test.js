import {spawn} from 'node:child_process'
import {readdir} from 'node:fs/promises'
import {Buffer} from 'node:buffer'
import {cwd} from 'node:process'

/** @type {Promise<{code: number, out: Buffer, err: Buffer}>} */
function lint(
    file,
) {
    const child = spawn('node', [
        'node_modules/.bin/eslint',
        '-c',
        'config.js',
        file,
    ], {
        stdio: ['pipe', 'pipe', 'pipe'],
        shell: true,
        env: {
            CI: 'true',
        },
    })

    const result = new Promise((res, rej) => {
        const outPromise = content(child.stdout)
        const errPromise = content(child.stderr)
        child.on('error', (e) => rej(e))
        child.on('exit', async (code) => {
            Promise.allSettled([outPromise, errPromise]).then(([out, err]) => {
                res({
                    code,
                    out: out.status === 'fulfilled' ? normalizeOutput(out.value.toString()) : String(out.reason),
                    err: err.status === 'fulfilled' ? err.value.toString() : String(err.reason),
                })
            })
        })
    })

    const kill = () => new Promise((res) => {
        if (child.exitCode === null) {
            child.on('close', () => res())
            child.kill()
        } else {
            res()
        }
    })

    child.stdin.end()

    return {child, result, kill}
}

/** @returns {Promise<Buffer>} */
function content(
    /** @type {NodeJS.ReadableStream} */
    stream,
) {
    return new Promise((res, rej) => {
        const chunks = []
        stream.on('data', c => chunks.push(c))
        stream.on('error', rej)
        stream.on('end', () => res(Buffer.concat(chunks)))
    })
}

function normalizeOutput(
    /** @type {string} */
    output,
) {
    return output.replace(cwd() + '/', '')
}

const cleanup = []
afterEach(async () => {
    const c = cleanup
    cleanup.splice(0)
    for (const f of c) {
        await f()
    }
})

for (const f of await readdir('examples')) {
    if (f === '.' || f === '..') {
        continue
    }

    test(`lint ${f}`, async () => {
        const {result, kill} = lint(`examples/${f}`)
        cleanup.push(kill)

        await expect(result).resolves.toMatchSnapshot()
    }, 30_000)
}

