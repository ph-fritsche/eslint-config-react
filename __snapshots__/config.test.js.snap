// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`lint javascript.js 1`] = `
{
  "code": 1,
  "err": "",
  "out": "
examples/javascript.js
  2:10  error  'foo' is defined but never used               no-unused-vars
  4:1   error  Expected indentation of 4 spaces but found 8  indent

✖ 2 problems (2 errors, 0 warnings)
  1 error and 0 warnings potentially fixable with the \`--fix\` option.

",
}
`;

exports[`lint javascriptxml.jsx 1`] = `
{
  "code": 1,
  "err": "",
  "out": "
examples/javascriptxml.jsx
  4:10  error  'foo' is defined but never used                                                                                                                                                                    no-unused-vars
  6:12  error  Avoid non-native interactive elements. If using native HTML is not possible, add an appropriate role and support for tabbing, mouse, keyboard, and touch inputs to an interactive content element  jsx-a11y/no-static-element-interactions

✖ 2 problems (2 errors, 0 warnings)

",
}
`;

exports[`lint typescript.ts 1`] = `
{
  "code": 1,
  "err": "",
  "out": "
examples/typescript.ts
  2:10  error  'foo' is defined but never used                                                                                                                                     @typescript-eslint/no-unused-vars
  4:5   error  Promises must be awaited, end with a call to .catch, end with a call to .then with a rejection handler or be explicitly marked as ignored with the \`void\` operator  @typescript-eslint/no-floating-promises

✖ 2 problems (2 errors, 0 warnings)

",
}
`;

exports[`lint typescriptxml.tsx 1`] = `
{
  "code": 1,
  "err": "",
  "out": "
examples/typescriptxml.tsx
  4:10  error  'foo' is defined but never used                                                                                                                                                                    @typescript-eslint/no-unused-vars
  6:12  error  Avoid non-native interactive elements. If using native HTML is not possible, add an appropriate role and support for tabbing, mouse, keyboard, and touch inputs to an interactive content element  jsx-a11y/no-static-element-interactions

✖ 2 problems (2 errors, 0 warnings)

",
}
`;
